### PR TITLE
[Coop] Mostly remove handles from Monitor.

### PR DESCRIPTION
**Pull request by jaykrell**
This is an alternative to https://github.com/mono/mono/pull/16980
or https://github.com/mono/mono/pull/16990.
It does *not* depend on
   https://github.com/mono/mono/pull/16987
or https://github.com/mono/mono/pull/17009
or similar.

It does *not* yet remove handles from Enter, due to Enter being intrinsic.
I still like the others but the PRs have been ignored.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Because compacting/moving GC?
I guess sgen does move?
Only when all references are from heap?
And so want a more significant record
of references from stack?

**Comment by jaykrell** [comment]
Maybe something in marshal-ilgen instead?
Something like https://github.com/mono/mono/pull/17009 ?

**Comment by jaykrell** [comment]
The CoreCLR pattern embodies the stack requirement in the name?
Harder for people to misuse?
Enforceable?

**Comment by jaykrell** [comment]
/cc @lambdageek 
(Granted a merge conflict, but not a significant one probably.)

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by akoeplinger** [comment]
@jaykrell FYI you can do `@monojenkins build monolite` now to just build monolit

**Comment by jaykrell** [comment]
@akoeplinger nice thank you! That should be much quicker. I recall it taking 30 minutes just to tar up stuff unnecessarily.

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
/cc @lambdageek
Previously was “ok”. Good enough?

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Old workflow just in case.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build monolite

**Comment by jaykrell** [comment]
@monojenkins build monolite

File: netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs [hunk 1]
```diff
@@ -31,14 +31,18 @@ public void Dispose ()
...
 		private bool WaitCore (int timeoutMs)
 		{
 			WaitEntry wait_entry = new WaitEntry ();
 			bool mutexLocked = false;
 			bool waitEntryLocked = false;
+			bool allowInterruption = false;
 
 			try {
-				Monitor.try_enter_with_atomic_var (mutex, Timeout.Infinite, false, ref mutexLocked);
+				Monitor.try_enter_with_atomic_var (ref mutex, Timeout.Infinite, allowInterruption, ref mutexLocked);
```
Comment (lambdageek) [inline netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs]: `ref mutex` here is a ref to a field in `this` (which is on the heap), so we don't actually have a guarantee that whatever `mutex` points to is pinned.

File: netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs [hunk 2]
```diff
@@ -60,9 +64,9 @@ private bool WaitCore (int timeoutMs)
 			}
 
 			try {
-				Monitor.try_enter_with_atomic_var (wait_entry.condition, Timeout.Infinite, false, ref waitEntryLocked);
+				Monitor.try_enter_with_atomic_var (ref wait_entry.condition, Timeout.Infinite, allowInterruption, ref waitEntryLocked);
```
Comment (lambdageek) [inline netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs]: this is pretty fragile - `wait_entry` happens to be a struct, so it's heap allocated, so `ref wait_entry.condition` is a pointer to the stack (and so whatever `condition` points to is pinned), but if `wait_entry` were a class, it wouldn't be.

File: netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs (hunk 1 above)
Comment (jaykrell) [reply netcore/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs]: I didn't realize this, but, really? It doesn't suffice?
Solution is to copy it to a local and ref that I guess, and then no need to answer prior question.

**Comment by akoeplinger** [comment]
The mono/mono repository is focused on stability and limiting code churn now so I don't think this change is worth doing here. If you think the change is important please reopen it in dotnet/runtime, thank you!

**Pull request closed without merging**